datatypes/src/raster/grid_array.rs

```diff
@@ -0,0 +1,393 @@
+        x as usize * stride_x
+    }
+
+    fn linear_space_index<I: Into<GridIdx<Self::IndexArray>>>(&self, index: I) -> Result<usize> {
+        let real_index = index.into();
+        ensure!(
+            self.bounding_box().contains(real_index),
+            error::GridIndexOutOfBounds {
+                index: Vec::from(real_index.0),
+                min_index: Vec::from(self.min_index().0),
+                max_index: Vec::from(self.max_index().0)
+            }
+        );
+
+        Ok(self.linear_space_index_unchecked(real_index))
+    }
+}
+
+impl GridBounds for ArrayShape1D {
+    type IndexArray = [isize; 1];
+
+    fn min_index(&self) -> GridIdx<Self::IndexArray> {
+        GridIdx::<[isize; 1]>::zero()
+    }
+
+    fn max_index(&self) -> GridIdx<Self::IndexArray> {
+        let [x_size] = self.shape_array;
+        GridIdx([x_size as isize]) - 1
+    }
+}
+
+impl GridSize for ArrayShape2D {
+    type ShapeArray = [usize; 2];
+
+    const NDIM: usize = 2;
+
+    fn axis_size(&self) -> Self::ShapeArray {
+        self.shape_array
+    }
+
+    fn number_of_elements(&self) -> usize {
+        let [a, b] = self.axis_size();
+        a * b
+    }
+}
+
+impl GridSpaceToLinearSpace for ArrayShape2D {
+    type IndexArray = [isize; 2];
+
+    fn strides(&self) -> Self::ShapeArray {
+        [self.axis_size_x(), 1]
+    }
+
+    fn linear_space_index_unchecked<I: Into<GridIdx<Self::IndexArray>>>(&self, index: I) -> usize {
+        let GridIdx([y, x]) = index.into();
+        let [stride_y, stride_x] = self.strides();
+        y as usize * stride_y + x as usize * stride_x
+    }
+
+    fn linear_space_index<I: Into<GridIdx<Self::IndexArray>>>(&self, index: I) -> Result<usize> {
+        let real_index = index.into();
+        ensure!(
+            self.bounding_box().contains(real_index),
+            error::GridIndexOutOfBounds {
+                index: Vec::from(real_index.0),
+                min_index: Vec::from(self.min_index().0),
+                max_index: Vec::from(self.max_index().0)
+            }
+        );
+        Ok(self.linear_space_index_unchecked(real_index))
+    }
+}
+
+impl GridBounds for ArrayShape2D {
+    type IndexArray = [isize; 2];
+
+    fn min_index(&self) -> GridIdx<Self::IndexArray> {
+        GridIdx::<[isize; 2]>::zero()
+    }
+
+    fn max_index(&self) -> GridIdx<Self::IndexArray> {
+        let [y_size, x_size] = self.shape_array;
+        GridIdx([y_size as isize, x_size as isize]) - 1
+    }
+}
+
+impl GridSize for ArrayShape3D {
+    type ShapeArray = [usize; 3];
+
+    const NDIM: usize = 3;
+
+    fn axis_size(&self) -> Self::ShapeArray {
+        self.shape_array
+    }
+
+    fn number_of_elements(&self) -> usize {
+        let [a, b, c] = self.axis_size();
+        a * b * c
+    }
+}
+
+impl GridSpaceToLinearSpace for ArrayShape3D {
+    type IndexArray = [isize; 3];
+
+    fn strides(&self) -> Self::ShapeArray {
+        [self.axis_size_y(), self.axis_size_x(), 1]
```

Comment:
shouldn't it be `[self.axis_size_y() * self.axis_size_x(), ...`?

Reply:
done